drake/examples/HolonomicDrive/HolonomicDrivePlant.m

```diff
@@ -0,0 +1,246 @@
+classdef HolonomicDrivePlant < SecondOrderSystem
+  properties (SetAccess = private, GetAccess = public)
+    m
+    I
+    wheels
+  end
+
+  methods
+    function obj = HolonomicDrivePlant(wheels, I, m)
+      % wheels: struct of
+      %   pos:      offset from center of mass
+      %   r:        wheel radius
+      %   driveDir: drive direction, a unit vector
+      %   slipDir:  slip direction. For mecanum wheels, this is not
+      %             orthogonal to the drive direction
+      %   omegaMax: motor free-running speed
+      %   tauMax:   motor stall-torque
+      % I: moment of inertia of entire robot about vertical axis
+      % m: total mass of robot
+      n = length(wheels);
+
+      obj = obj@SecondOrderSystem(3, n, true);
+
+      obj.I = I;
+      obj.m = m;
+      obj.wheels = wheels;
+
+      obj = setInputLimits(obj,-1,1);
+
+      obj = setInputFrame(obj,CoordinateFrame('HolonomicInput',n,'u',...
+        arrayfun(@(i) sprintf('u%d', i), 1:n, 'Unif', false)));
+      obj = setStateFrame(obj,CoordinateFrame('HolonomicState',6,'x',...
+        {'x','y','theta', 'x_dot','y_dot', 'theta_dot'}));
+      obj = setOutputFrame(obj,obj.getStateFrame);
+    end
+
+    function speeds = rotorSpeeds(obj, vel, omega)
+      % get the motor rotor speeds needed to give the robot a certain
+      % velocity in body space
+      i = 1;
+      speeds = zeros(length(obj.wheels), 1)*vel(1); % to help TaylorVar
+      for wheel = obj.wheels
+        % r cross omega
+        rotv = omega * [-wheel.pos(2); wheel.pos(1)];
+        totalv = vel + rotv;
+
+        % decompose into powered and unpowered direction
+        vwheel = [wheel.driveDir wheel.slipDir] \ totalv;
+
+        speeds(i) = vwheel(1) / wheel.r;
+        i = i + 1;
+      end
+    end
+
+    function [qdd, qdddx] = sodynamics(obj,t,q,qd,u)
+      n = length(obj.wheels);
+
+      if nargout > 1
+        qdddx = zeros(3, 7 + n);
+        % dynamics are linear in each wheel.
+        for i = 1:n
+          wheel = obj.wheels(i);
+          qdddxi = oneWheelGradients(obj.m, obj.I, wheel, t, q, qd, u(i));
+          qdddx(:,1:7) = qdddx(:,1:7) + qdddxi(:,1:7);
+
+          % qdd / du_i
+          qdddx(:,7+i) = qdddxi(:,8);
+        end
+      end
+
+      theta = q(3);
+      rotation = [[cos(theta); sin(theta)] [-sin(theta); cos(theta)]];

```

Comment:
use rotmat from drake/util (nitpick, but should always prefer a method that's been tested by being used everywhere).